---
title: "LEUK45_Preprocessing"
author: "Aditi Vedi"
date: "2023/11/24"
output: html_document
---

```{r}
library(tidyverse)
library(ggpubr)
library(plater)
library(reshape2)
```

## Read in index data
```{r}
Index_P1 <- read_csv("Input_data/Index_data/Index_P1.csv") %>% 
  mutate(Plate = 1)
Index_P2 <- read_csv("Input_data/Index_data/Index_P2.csv")%>% 
  mutate(Plate = 2)
Index_P3 <- read_csv("Input_data/Index_data/Index_P3.csv")%>% 
  mutate(Plate = 3)
Index_P4 <- read_csv("Input_data/Index_data/Index_P4.csv")%>% 
  mutate(Plate = 4)
Index_P5 <- read_csv("Input_data/Index_data/Index_P5.csv")%>% 
  mutate(Plate = 5)
Index_P6 <- read_csv("Input_data/Index_data/Index_P6.csv")%>% 
  mutate(Plate = 6)
Index_P7 <- read_csv("Input_data/Index_data/Index_P7.csv")%>% 
  mutate(Plate = 7)
Index_P8 <- read_csv("Input_data/Index_data/Index_P8.csv")%>% 
  mutate(Plate = 8)
Index_P9 <- read_csv("Input_data/Index_data/Index_P9.csv")%>% 
  mutate(Plate = 9)
Index_P10 <- read_csv("Input_data/Index_data/Index_P10.csv")%>% 
  mutate(Plate = 10)

LEUK45_Index <- rbind(Index_P1, Index_P2, Index_P3, Index_P4, Index_P5, Index_P6, Index_P7, Index_P8, Index_P9, Index_P10) %>% 
  select(contains("Median") & starts_with("Cells"), Well, Plate) %>% 
  unite(Sort_id, Plate, Well) %>% 
  mutate(Cell_seeded = "HSC")

colnames(LEUK45_Index)

colnames(LEUK45_Index)[1:ncol(LEUK45_Index)] <- str_remove_all(colnames(LEUK45_Index)[1:ncol(LEUK45_Index)], pattern = "Cells | Median") #Remove certain words from the column names
colnames(LEUK45_Index)[1:ncol(LEUK45_Index)] <- word(colnames(LEUK45_Index)[1:ncol(LEUK45_Index)], 1) #Remove certain words from the column names
colnames(LEUK45_Index) <- str_replace(colnames(LEUK45_Index), pattern = "-", replacement = "_")


colnames(LEUK45_Index)

LEUK45_Index %>% count(Cell_seeded)

write_csv(LEUK45_Index, "Processed_data/Index_data/LEUK45_Index.csv")
```

```{r}
# Read in collated Index data

Index <- read_csv("Processed_data/Index_data/LEUK45_Index.csv")
```
```{r}

```

## Read in Phenotyping data

```{r}
file_path <- "Input_data/Phenotype_data/Harvest_plan_LEUK45_FACS.csv"

check_plater_format(file_path)

plate <- read_plate(file = file_path, well_ids_column = 'Wells')
plate$Wells <- str_remove(plate$Wells, '(?<=[A-Z])0*(?=[0-9])')

Plate_map <- melt(plate, id = "Wells") %>% 
  filter(!is.na(value)) %>%
  mutate(Plate = as.character(variable), 
         Plate = parse_number(Plate)) %>% 
  unite(Plate_well, Plate, Wells) %>% 
  rename(Sort_id = value) %>% 
  select(-variable)

write_csv(Plate_map, "Processed_data/Phenotype_data/LEUK45_Plate_map.csv")
```

```{r}
Harvested_colonies <- left_join(Plate_map, Index, by = "Sort_id")

Harvested_colonies %>% count(Cell_seeded)
```
```{r}

```

## Read in FACS data
```{r}
FACS <- read_csv("Input_data/Phenotype_data/20231124_LEUK45_ARCH4_TrueFalse.csv") %>% 
  filter(TrueFalse == "TRUE") %>% 
  separate(Plate_well, into = c("Plate", "Well"), sep = "_")

FACS$Well <- str_remove(FACS$Well, '(?<=[A-Z])0*(?=[0-9])')

FACS <- FACS %>% 
  unite(Plate_well, Plate, Well)
```
```{r}
# Join FACS with sort_id from plate map

True_colonies <- left_join(FACS, Harvested_colonies, by = "Plate_well") %>% 
  filter(TrueFalse == "TRUE")

True_colonies %>% count(Cell_seeded)

write_csv(True_colonies, "Processed_data/Phenotype_data/LEUK45_True_colonies.csv")
```
```{r}

```

## Read in current seq VAF data

```{r}
Run <- read_csv("Input_data/Genotype_data/run_statistics_1.csv") %>% 
  filter(str_detect(...1, "BM")) %>% 
  separate(...1, into = c("LEUK", "Source", "Plate", "Well", "Sample_id")) %>% 
  mutate(Plate = str_remove(Plate, "P")) %>% 
  unite(Plate_well, Plate, Well) %>% 
  mutate(Exp = "LEUK45") %>% 
  select(Source, Plate_well, Sample_id, Exp, `2`) %>% 
  rename(Depth = `2`)
```
### Varscan

```{r}
Varscan <- read_csv("Input_data/Genotype_data/total_varscan.csv") %>% 
  mutate(Start = as.numeric(Start))
```
```{r}
# Keeping only samples from our run
# looking at DNMT3A R882 (i.e. Start position 25457243)

Varscan <- Varscan %>% 
  filter(str_detect(sample_name, "Leuk_BM"), Gene.refGene == "DNMT3A", Start == 25457243) %>% 
  select(Chr, Start, End, Ref, sample_name, Depth, VAF, Gene.refGene) %>% 
  rename(VAF.varscan = VAF, Depth.varscan = Depth) %>% 
  separate(sample_name, into = c(NA, "Source", "Plate", "Well", "Sample_id"), sep = "_") %>% 
  mutate(Plate = str_remove(Plate, "P")) %>% 
  unite(Plate_well, Plate, Well) %>% 
  select(-Source)
```
### Mutect
```{r}
Mutect <- read_csv("Input_data/Genotype_data/total_mutect.csv") %>% 
  mutate(Start = as.numeric(Start))
```
```{r}
# Keeping only samples from our run
# looking at DNMT3A R882 (i.e. Start position 25457243)

Mutect <- Mutect %>% 
  filter(str_detect(sample_name, "Leuk_BM"), Gene.refGene == "DNMT3A", Start == 25457243) %>% 
  select(Chr, Start, End, Ref, sample_name, Depth, VAF, Gene.refGene) %>% 
  rename(VAF.mutect = VAF, Depth.mutect = Depth) %>% 
  separate(sample_name, into = c(NA, "Source", "Plate", "Well", "Sample_id"), sep = "_") %>% 
  mutate(Plate = str_remove(Plate, "P")) %>% 
  unite(Plate_well, Plate, Well) %>% 
  select(-Source)
```
### Platypus
```{r}
Platypus <- read_csv("Input_data/Genotype_data/total_platypus.csv") %>% 
  mutate(Start = as.numeric(Start))
```
```{r}
# Keeping only samples from our run
# looking at DNMT3A R882 (i.e. Start position 25457243)

Platypus <- Platypus %>% 
  filter(str_detect(sample_name, "Leuk_BM"), Gene.refGene == "DNMT3A", Start == 25457243) %>% 
  select(Chr, Start, End, Ref, sample_name, Depth, VAF, Gene.refGene) %>% 
  rename(VAF.platypus = VAF, Depth.platypus = Depth) %>% 
  separate(sample_name, into = c(NA, "Source", "Plate", "Well", "Sample_id"), sep = "_") %>% 
  mutate(Plate = str_remove(Plate, "P")) %>% 
  unite(Plate_well, Plate, Well) %>% 
  select(-Source)
```
```{r}
# Select only relevant columns

Varscan <- Varscan %>% 
  select(Chr, Start, Sample_id, Depth.varscan, VAF.varscan, Gene.refGene)

Mutect <- Mutect %>% 
  select(Sample_id, Depth.mutect, VAF.mutect)

Platypus <- Platypus %>% 
  select(Sample_id, Depth.platypus, VAF.platypus)
```
```{r}
# Combine all above data into single file containing depth from run stats and VAF from 3 variant callers

VAF <- left_join(Run, Varscan, by = "Sample_id")
VAF <- left_join(VAF, Mutect, by = "Sample_id")
VAF <- left_join(VAF, Platypus, by = "Sample_id")

write_csv(VAF, "Processed_data/Genotype_data/VAF.csv")
```

```{r}
VAF <- VAF %>% 
  arrange(Depth)
```

```{r}
# filter out the colonies with low depth. i.e. keep colonies above the lowest depth that has at least 2 variant callers that call a colony mutant

Temp_VAF <- VAF[ rowSums(VAF[c('VAF.varscan',	'VAF.mutect',	'VAF.platypus')] > 0.02 , na.rm=TRUE ) >= 2,]
Depth_cutoff <- min(Temp_VAF$Depth)

VAF <- VAF %>% 
  filter(Depth>=Depth_cutoff)

```
```{r}
# Calculate mean VAF based on the 3 variant callers

VAF <- VAF %>% 
    rowwise() %>%
  mutate(mean_VAF = mean(c(VAF.varscan, VAF.mutect, VAF.platypus), na.rm = TRUE)) %>%
  ungroup()
```

```{r}
# If any of the VAFs' mean has been returned as NA (i.e. all variant callers were NA) then give it VAF = 0 ie WT
VAF$mean_VAF[is.nan(VAF$mean_VAF)]<-0
```
```{r}

```

### Assign Mut/WT to VAFs
```{r}
VAF <- VAF %>% 
  mutate (DNMT3A = case_when(mean_VAF < 0.02 ~ "WT",
                                 mean_VAF >= 0.02 ~ "MUT"),
          DNMT3A = fct_relevel(DNMT3A, c("WT", "MUT")))

VAF %>% count(DNMT3A)

write_csv(VAF, "Processed_data/Genotype_data/LEUK45_final_VAF.csv")
```
```{r}

```

## Join VAF with True colonies
```{r}
VAF <- left_join(Plate_map, VAF, by = "Plate_well")

Final_data <- left_join(VAF, True_colonies, by = "Sort_id") %>% 
  filter(!is.na(fcs.name))

Final_data %>% count(DNMT3A)
```

```{r}

```

```{r}
dim(Final_data)
```

```{r}

```

```{r}
# clean up and save final data

names(Final_data)[names(Final_data) == 'Plate_well.x'] <- 'Plate_well'

drop = c('Plate_well.y')

Final_data <- Final_data[,!(names(Final_data) %in% drop)]

dim(Final_data)
```

```{r}

```

```{r}
filename_prefix <- format(as.Date(Sys.time()), "%Y%m%d")

outname <- paste0("Processed_data/",filename_prefix,"_LEUK45_Final_data.csv")

write_csv( Final_data,  outname)

print(outname)
```

```{r}

```

```{r}
sessionInfo()
```



